### PR TITLE
Add IPEX-LLM with GPU

### PR DESCRIPTION
**Pull request by ivy-lv11**
- **Description**: [`ipex-llm`](https://github.com/intel-analytics/ipex-llm) is a PyTorch library for running LLM on Intel CPU and GPU. This PR adds GPU support to the IpexLLM llm integration.
- **Fixes # (issue)**: N/A
- **New Package?**: No
- **Examples**
     - added a few more examples
     - added `-d` option for all examples to choose device.
     - add a new jupyter notebook doc for GPU. 
- **Tests**: N/A

### CODE REVIEW

File: .github/workflows/publish_sub_package.yml [hunk 1]
```diff
@@ -4,14 +4,18 @@ on:
   push:
     branches:
       - main
+    pull_request:
+      branches: [main]
```
Comment (Oscilloscope98) [inline .github/workflows/publish_sub_package.yml]: Seems like the `publish_sub_packages` workflow has not been successfully triggered. Maybe change to
branches: [main, ipex-llm-llm-gpu]

and have a try again

Comment (ivy-lv11) [reply .github/workflows/publish_sub_package.yml]: Already added

File: .github/workflows/publish_sub_package.yml [hunk 2]
```diff
@@ -4,14 +4,19 @@ on:
   push:
     branches:
       - main
+    pull_request:
```
Comment (Oscilloscope98) [inline .github/workflows/publish_sub_package.yml]: `pull_request` should at same level with `push`

Comment (ivy-lv11) [reply .github/workflows/publish_sub_package.yml]: Fixed

**Comment by shane-huang** [comment]
- add a jupyter notebook for gpu like this https://github.com/run-llama/llama_index/pull/13097/files#diff-4de70abde948df04dbda0c1a6cc564e1bf27c5481665939db1f014df156e6679
- add a new example which works for both CPU and GPU, migrate this example to use the new interface https://github.com/intel-analytics/ipex-llm/blob/main/python/llm/example/GPU/LlamaIndex/rag.py. 
- update the more_data_types to use GPU, but need to test other data types which only works on GPU, etc. fp4

File: llama-index-integrations/llms/llama-index-llms-ipex-llm/examples/more_data_type.py [hunk 3]
```diff
@@ -33,11 +33,20 @@ def completion_to_prompt(completion):
         choices=["sym_int4", "asym_int4", "sym_int5", "asym_int5", "sym_int8"],
```
Comment (shane-huang) [inline llama-index-integrations/llms/llama-index-llms-ipex-llm/examples/more_data_type.py]: update the choices to and add GPU related data_types. For a full list of datatypes we can support, refer to https://ipex-llm.readthedocs.io/en/latest/doc/PythonAPI/LLM/transformers.html#automodelforcausallm `load_in_lowbit` param api doc

**Comment by ivy-lv11** [comment]
> * add a jupyter notebook for gpu like this https://github.com/run-llama/llama_index/pull/13097/files#diff-4de70abde948df04dbda0c1a6cc564e1bf27c5481665939db1f014df156e6679
> * add a new example which works for both CPU and GPU, migrate this example to use the new interface https://github.com/intel-analytics/ipex-llm/blob/main/python/llm/example/GPU/LlamaIndex/rag.py.
> * update the more_data_types to use GPU, but need to test other data types which only works on GPU, etc. fp4

Various data types were tested, including: fp4, fp8, fp16, bf16, nf3, nf4, fp8_e4m3, fp8_e5m2. All generate normally.

File: llama-index-integrations/llms/llama-index-llms-ipex-llm/examples/README.md [hunk 4]
```diff
@@ -12,17 +12,92 @@ Install `llama-index-llms-ipex-llm`. This will also install `ipex-llm` and its d
...
+- **Data Preparation**: Download the Llama2 paper and save it as `data/llama2.pdf`, which serves as the default source file for retrieval.
+  ```bash
+  mkdir data
+  wget --user-agent "Mozilla" "https://arxiv.org/pdf/2307.09288.pdf" -O "data/llama2.pdf"
+  ```
+
+The example [rag.py](./rag.py) shows how to use RAG pipeline. Run the example as following:
+
+```bash
+python rag.py -m <path_to_model> -q <question> -u <vector_db_username> -p <vector_db_password> -e <path_to_embedding_model> -n <num_token> -t <path_to_tokenizer> -x <device>
```
Comment (shane-huang) [inline llama-index-integrations/llms/llama-index-llms-ipex-llm/examples/README.md]: we'd better to use the same option `-d` for device for all examples。

File: llama-index-integrations/llms/llama-index-llms-ipex-llm/examples/more_data_type.py [hunk 5]
```diff
@@ -33,11 +33,20 @@ def completion_to_prompt(completion):
...
         help="The quantization type the model will convert to.",
     )
+    parser.add_argument(
+        "--device",
+        "-d",
+        type=str,
+        default="xpu",
+        choices=["cpu", "xpu", "auto"],
```
Comment (shane-huang) [inline llama-index-integrations/llms/llama-index-llms-ipex-llm/examples/more_data_type.py]: remove the auto option, we don't support this.

**Comment by shane-huang** [comment]
- remove `rag.py` and we shall add it to the next PR. 
- write a simple inference example `basic.py` just like in jupyter and update example/README
- update jupyter just like the embedding jupyter 
- uniform the arguments d, -q, -m params for all examples. 
- support only cpu/xpu and report error for others

**Comment by shane-huang** [comment]
also fix the test errors

**Comment by shane-huang** [comment]
> > * add a jupyter notebook for gpu like this https://github.com/run-llama/llama_index/pull/13097/files#diff-4de70abde948df04dbda0c1a6cc564e1bf27c5481665939db1f014df156e6679
> > * add a new example which works for both CPU and GPU, migrate this example to use the new interface https://github.com/intel-analytics/ipex-llm/blob/main/python/llm/example/GPU/LlamaIndex/rag.py.
> > * update the more_data_types to use GPU, but need to test other data types which only works on GPU, etc. fp4
> 
> Various data types were tested, including: fp4, fp8, fp16, bf16, nf3, nf4, fp8_e4m3, fp8_e5m2. All generate normally.

add those options to example choices

File: llama-index-integrations/llms/llama-index-llms-ipex-llm/examples/more_data_type.py (hunk 5 above)
Comment (ivy-lv11) [reply llama-index-integrations/llms/llama-index-llms-ipex-llm/examples/more_data_type.py]: removed.

File: docs/docs/examples/llm/ipex_llm_gpu.ipynb [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+    "\n",
+    "> **Note**\n",
+    ">\n",
+    "> For the first time that each model runs on Intel iGPU/Intel Arc A300-Series or Pro A60, it may take several minutes to compile.\n",
+    ">\n",
+    "> For other GPU type, please refer to [here](https://ipex-llm.readthedocs.io/en/latest/doc/LLM/Overview/install_gpu.html#runtime-configuration) for Windows users, and  [here](https://ipex-llm.readthedocs.io/en/latest/doc/LLM/Overview/install_gpu.html#id5) for Linux users.\n",
+    "\n",
+    "## `IpexLLM`\n",
+    "\n",
+    "Setting `device_map=\"xpu\"` when initializing `IpexLLM` will put the embedding model on Intel GPU and benefit from IPEX-LLM optimizations:\n",
```
Comment (shane-huang) [inline docs/docs/examples/llm/ipex_llm_gpu.ipynb]: change this line to use the descriptions in llm jupyter doc. 
add the descriptions to explain prompts like in llm jupyter doc

File: llama-index-integrations/llms/llama-index-llms-ipex-llm/examples/basic.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        help="The huggingface repo id for the LLM model to be downloaded"
+        ", or the path to the huggingface checkpoint folder",
+    )
+    parser.add_argument(
+        "--device",
+        "-d",
+        type=str,
+        default="cpu",
+        choices=["cpu", "xpu"],
+        help="The device (Intel CPU or Intel GPU) the embedding model runs on",
```
Comment (shane-huang) [inline llama-index-integrations/llms/llama-index-llms-ipex-llm/examples/basic.py]: embedding model -> llm model

File: llama-index-integrations/llms/llama-index-llms-ipex-llm/examples/low_bit.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+        help="The huggingface repo id for the LLM model to be downloaded"
+        ", or the path to the huggingface checkpoint folder",
+    )
+    parser.add_argument(
+        "--device",
+        "-d",
+        type=str,
+        default="cpu",
+        choices=["cpu", "xpu"],
+        help="The device (Intel CPU or Intel GPU) the embedding model runs on",
```
Comment (shane-huang) [inline llama-index-integrations/llms/llama-index-llms-ipex-llm/examples/low_bit.py]: embedding model -> llm model

File: llama-index-integrations/llms/llama-index-llms-ipex-llm/llama_index/llms/ipex_llm/base.py [hunk 9]
```diff
@@ -197,7 +197,11 @@ def __init__(
             self._model = self._load_model(
                 low_bit_model, load_in_4bit, load_in_low_bit, model_name, model_kwargs
             )
-
+        if device_map not in ["cpu", "xpu"]:
+            raise ValueError(
+                "IpexLLMEmbedding currently only supports device to be 'cpu' or 'xpu', "
```
Comment (shane-huang) [inline llama-index-integrations/llms/llama-index-llms-ipex-llm/llama_index/llms/ipex_llm/base.py]: IpexLLM

File: llama-index-integrations/llms/llama-index-llms-ipex-llm/llama_index/llms/ipex_llm/base.py [hunk 10]
```diff
@@ -288,7 +292,7 @@ def from_model_id(
         load_in_low_bit: Optional[str] = None,
         model: Optional[Any] = None,
         tokenizer: Optional[Any] = None,
-        device_map: Optional[str] = "auto",
+        device_map: Optional[str] = "cpu",
```
Comment (shane-huang) [inline llama-index-integrations/llms/llama-index-llms-ipex-llm/llama_index/llms/ipex_llm/base.py]: use Literal

File: llama-index-integrations/llms/llama-index-llms-ipex-llm/llama_index/llms/ipex_llm/base.py [hunk 11]
```diff
@@ -334,7 +338,7 @@ def from_model_id_low_bit(
         model_name: str = DEFAULT_HUGGINGFACE_MODEL,
         model: Optional[Any] = None,
         tokenizer: Optional[Any] = None,
-        device_map: Optional[str] = "auto",
+        device_map: Optional[str] = "cpu",
```
Comment (shane-huang) [inline llama-index-integrations/llms/llama-index-llms-ipex-llm/llama_index/llms/ipex_llm/base.py]: use Literal

File: llama-index-integrations/llms/llama-index-llms-ipex-llm/examples/low_bit.py [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+        tokenizer_name=model_name,
+        context_window=512,
+        max_new_tokens=128,
+        generate_kwargs={"do_sample": False},
+        completion_to_prompt=completion_to_prompt,
+        messages_to_prompt=messages_to_prompt,
+        device_map=device,
+    )
+
+    saved_lowbit_model_path = "./zephyr-7b-alpha-low-bit"  # path to save low-bit model
```
Comment (shane-huang) [inline llama-index-integrations/llms/llama-index-llms-ipex-llm/examples/low_bit.py]: change saved path to an argument, and name it as `./saved_lowbit_model`

File: llama-index-integrations/llms/llama-index-llms-ipex-llm/examples/more_data_type.py [hunk 13]
```diff
@@ -30,14 +30,45 @@ def completion_to_prompt(completion):
...
         default="asym_int4",
-        choices=["sym_int4", "asym_int4", "sym_int5", "asym_int5", "sym_int8"],
+        choices=[
+            "sym_int4",
+            "asym_int4",
+            "sym_int5",
+            "asym_int5",
+            "sym_int8",
+            "fp4",
+            "fp8",
```
Comment (shane-huang) [inline llama-index-integrations/llms/llama-index-llms-ipex-llm/examples/more_data_type.py]: add some comments to tell user which options are only available on GPU

File: llama-index-integrations/llms/llama-index-llms-ipex-llm/examples/README.md [hunk 14]
```diff
@@ -12,16 +12,46 @@ Install `llama-index-llms-ipex-llm`. This will also install `ipex-llm` and its d
...
 
+### Basic Example
+
+The example [basic.py](./basic.py) shows how to run `IpexLLM` on Intel CPU or GPU and conduct tasks such as text completion. Run the example as following:
+
+```bash
+python basic.py -m <path_to_model> -d <cpu_or_xpu> -q <query_to_LLM>
+```
+
+> Please note that in this example we'll use [HuggingFaceH4/zephyr-7b-alpha](https://huggingface.co/HuggingFaceH4/zephyr-7b-alpha) model for demonstration. It requires updating `transformers` and `tokenizers` packages.
```
Comment (shane-huang) [inline llama-index-integrations/llms/llama-index-llms-ipex-llm/examples/README.md:33]: lowbit also uses zephyr, put this update also in the low bit exmaple

**Comment by Oscilloscope98** [comment]
Why we use langchain description here :)?

**Pull request merged**